Feature: Extend PartsOf to mock non-virtual methods implementing an i…

marcoregueira opened the pull request:
Hi 

I was in need of a feature allowing PartsOf to mock non-virtual methods implementing an interface, just the same as requested in #625.

I decided to tinker a bit to see If there was any way of doing it. Solution came out faster than expected and here are the results. I thought you might find it useful enough to include it.

In short, Castle already provides this feature and the task was, mostly, opening a way to expose it without breaking anything.

How to use:
~~`var substitute = Substitute.ForPartsOf<ISomeInterface,SomeImplementation>(argsList);`~~
`var substitute = Substitute.ForTypeForwardingTo<ISomeInterface,SomeImplementation>(argsList);`

Tests are included, of course, to see it in action.

#### Obvious limitations: 

Overriding virtual methods effectively replaces its implementation both for internal and external calls. With this implementation Nsubstitute will only intercept calls made by client classes using the interface. Calls made from inside the object itself to its own method will hit the actual implementation.

#### Possible improvements

The next logical improvement is to provide mocking for already created objects conforming to an interface. Something like this:

```
ISomeInterface  someObject = new SomeClass();
ISomeInterface substitute = Substitute.ForInstanceForwardingTo<ISomeInterface>(someObject);
```

Please kindly take a look and give your feedback! Thanks!

marcoregueira commented:
Nice feature to test the examples in the documentation. 😄 There was an use case in the docs that was not in the code, in the file `2010-01-02-creating-a-substitute.markdown`.


![image](https://user-images.githubusercontent.com/2399984/180312082-f3e74099-8973-486b-8bca-dabfcbfc4612.png)

There was a regression because the new ForPartsOf is based on For<>, and there was a new check to ensure that the class implemented the interface. It did not break any of the existing tests but it did with those on the documentation.

I have solved it and copied the test into the solution.

Now `ForPartsOf<ITestInterface, TestClass>` will behave exactly as `For<>` if `TestClass` doesn't implement the interface. If you feel the check is necessary, I think I will just add it to the implementation of `ForPastOf `as the first step. It would be more difficult to check this at a later step.

marcoregueira commented:
I've found that the use case mentioned in my previous comment and taken from the docs was not working ok for all scenarios.

In the case of partial substitutes, the substitute can't extend the base class, just the interface, since we are substituting non-virtual methods. I think this is an acceptable limitation. In my previous implementation, regretfully, this was happening for all other substitutes. While I believe it is an edge case, it would have meant a breaking change.

The new implementation involves passing a parameter around to know when we are creating a partial substitute or not. I find this a bit inelegant, but it causes less friction with the current code.

I've removed the previous commits and replaced them with a squashed version.


Relevant tests are in
`NSubstitute.Acceptance.Specs\SubbingForConcreteTypesAndMultipleInterfaces.cs`
and
`NSubstitute.Acceptance.Specs\PartialSubs.cs`

dtchepak reviewed (commented):
Sorry for the delay in reviewing this. 😓 🙇 

This is a really nice PR, thanks! I'm wondering if, to be extra careful, we should leave the existing `Substitute.ForPartsOf` unchanged, and expose this functionality through `Substitute.ForTypeForwardingTo()` or similar? That way changes to the types passed will not result in significant changes to how the resulting substitute behaves. WDYT?

marcoregueira commented:
Hi!
Thank you for taking the time to review the PR. I'm glad you like it. I
will make the amendments you suggested.

marcoregueira commented:
Hi again

I've added the changes you suggested:

* Renamed the method to `ForTypeForwardingTo` and moved all tests, accordingly, to a new class named `TypeForwarding`. All changes in `PartialSubs` have been removed.

* Created new specific exceptions a shown in the following image. 

Please, kindly let me know if you find something else to improve.

![image](https://user-images.githubusercontent.com/2399984/198825196-50147594-c5f9-43e2-be27-8cf7a5f53564.png)

johnc-ftl commented:
I would love to have this integrated in NSubstitute... @marcoregueira  - thank you for your work on this!

johncrim commented:
I would love to have this integrated in NSubstitute... @marcoregueira  - thank you for your work on this!

NidhalWers commented:
Thanks @marcoregueira for your solution, it's something I look forward to have in NSubstitute.
@dtchepak Is there any blocker to integrate this ?

dtchepak commented:
> Thanks @marcoregueira for your solution, it's something I look forward to have in NSubstitute. @dtchepak Is there any blocker to integrate this ?

Sorry, this completely slipped off my radar. 😞 🙇 

Before I merged I wanted to add some tests to demonstrate equivalence between substitutes created using this method vs. `ForPartsOf` and manually forwarding to the concrete type for the standard NSub features.

If anyone has time to look at that I would greatly appreciate it, otherwise thanks a lot for the reminder and I'll try to look into this again shortly.

marcoregueira commented:
@dtchepak,

Thank you for your review.

I've upgraded the code to the latest version and made the necessary updates to the PR. To verify the new feature, please refer to the tests in `TypeForwarding.cs`.

dtchepak commented on src/NSubstitute/Substitute.cs:
```diff
@@ -89,4 +90,19 @@ public static T ForPartsOf<T>(params object[] constructorArguments)
         return (T)substituteFactory.CreatePartial([typeof(T)], constructorArguments);
     }
+
+        public static T ForTypeForwardingTo<T,TClass>(params object[] constructorArguments)
```
**chore:** could you please add doc comments here (see `ForPartsOf` and other public methods for examples).

dtchepak commented on src/NSubstitute/Substitute.cs:
```diff
@@ -89,4 +90,19 @@ public static T ForPartsOf<T>(params object[] constructorArguments)
+            return (T)substituteFactory.CreatePartial(new[] { typeof(T), typeof(TClass) }, constructorArguments);
+        }
+
+        //public static T ForTypeForwardingTo<T, T2, T3>(params object[] constructorArguments)
+        //    where T : class
+        //{
+        //    var substituteFactory = SubstitutionContext.Current.SubstituteFactory;
+        //    return (T)substituteFactory.CreatePartial(new[] { typeof(T), typeof(TClass) }, constructorArguments);
+        //}
```
**chore (nitpick):** please remove commented out code.

dtchepak commented on src/NSubstitute/Substitute.cs:
```diff
@@ -89,4 +90,19 @@ public static T ForPartsOf<T>(params object[] constructorArguments)
         var substituteFactory = SubstitutionContext.Current.SubstituteFactory;
         return (T)substituteFactory.CreatePartial([typeof(T)], constructorArguments);
     }
+
+        public static T ForTypeForwardingTo<T,TClass>(params object[] constructorArguments)
+            where T : class
+        {
+            var substituteFactory = SubstitutionContext.Current.SubstituteFactory;
+            return (T)substituteFactory.CreatePartial(new[] { typeof(T), typeof(TClass) }, constructorArguments);
+        }
+
```
**question:** not sure if it is just the diff view but the code indenting looks non-standard here?

dtchepak commented on src/NSubstitute/Proxies/ProxyFactory.cs:
```diff
@@ -1,15 +1,18 @@
+
+
```
**nitpick:** additional blank lines not required

dtchepak commented on src/NSubstitute/Core/SubstituteFactory.cs:
```diff
@@ -1,4 +1,6 @@
+
```
**question (nitpick):** several of these changed files have a blank line to added to start. Is this an editorconfig thing or accidental?

dtchepak reviewed (commented):
Looks great. Couple of very minor nitpicks to resolve then I'll merge 👍

marcoregueira commented:
Hi @dtchepak

I found some time to address the issues you raised. Thank you for reviewing them.

I've included the documentation section you requested. Please let me know if the wording needs further adjustment.

Feel free to let me know if you need any more changes!


Changes:

- Updated namespaces and adjusted indentation to align with the new file structure.
- Fixed changes in the includes section.
- Simplified collection initializations using collection expressions.
- Addressed primary constructor warnings.
- Removed unnecessary comments.
- Added documentation coments.

dtchepak commented on src/NSubstitute/Exceptions/TypeForwardingException.cs:
```diff
@@ -0,0 +1,21 @@
+{
+}
+
+public sealed class CanNotForwardCallsToClassNotImplementingInterfaceException(Type type) : TypeForwardingException(DescribeProblem(type))
+{
+    private static string DescribeProblem(Type type)
+    {
+        return string.Format("The provided class '{0}' doesn't implement all requested interfaces. ", type.Name);
+    }
```
**suggestion (non-blocking):** for a future change, I think it would be helpful for this to explain which of the requested interfaces has not been implemented. If we see "Class 'A' does not implement all requested interfaces. To support call forwarding 'A' must implement 'B', 'C', and 'D'." then it gives us a good hint on how to fix the exception.

dtchepak commented on src/NSubstitute/Exceptions/TypeForwardingException.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    private static string DescribeProblem(Type type)
+    {
+        return string.Format("The provided class '{0}' is abstract. ", type.Name);
```
**nitpick (non-blocking):** trailing space: `"...abstract . "` vs `"...abstract ."`..

dtchepak commented on src/NSubstitute/Proxies/CastleDynamicProxy/CastleInvocationMapper.cs:
```diff
@@ -10,8 +10,7 @@ public virtual ICall Map(IInvocation castleInvocation)
             castleInvocation.MethodInvocationTarget.IsVirtual &&
-            !castleInvocation.MethodInvocationTarget.IsAbstract &&
-            !castleInvocation.MethodInvocationTarget.IsFinal)
+            !castleInvocation.MethodInvocationTarget.IsAbstract)
```
**question:** dropping "not final" constraint here is required to support call forwarding?

dtchepak commented on src/NSubstitute/Substitute.cs:
```diff
@@ -89,4 +89,24 @@ public static T ForPartsOf<T>(params object[] constructorArguments)
+    /// <summary>
+    /// Creates a proxy for a class that implements an interface, forwarding methods and properties to an instance of the class, effectively mimicking a real instance.
+    /// Both the interface and the class must be provided as parameters.
+    /// The proxy will log calls made to the interface members and delegate them to an instance of the class. Specific members can be substituted 
```
**nitpick (non-blocking):** "will log calls" -> "will intercept calls"

dtchepak commented on src/NSubstitute/Substitute.cs:
```diff
@@ -89,4 +89,24 @@ public static T ForPartsOf<T>(params object[] constructorArguments)
         return (T)substituteFactory.CreatePartial([typeof(T)], constructorArguments);
     }
+
+    /// <summary>
+    /// Creates a proxy for a class that implements an interface, forwarding methods and properties to an instance of the class, effectively mimicking a real instance.
```
**praise:** great code docs here, thanks! 🙇

dtchepak commented:
Thanks @marcoregueira !
